### PR TITLE
solution1: add solution for exercise 1

### PR DESCRIPTION
**Pull request by kopiczko**
# WIP

### CODE REVIEW

**Comment by kopiczko** [comment]
@rossf7 this is still WIP, but you can check my approach. Any criticism welcome.

File: solution1/crd.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# 
+# NOTE: This file is not used in the solution this serves an example how the yaml
+# file looks like. Content of this file is copied crdJson variable in JSON
+# format.
```
Comment (rossf7) [inline solution1/crd.yaml]: Having the yaml file is nice <3 Nitpick but I'd rewrite the comment like this.

This file is not used in the solution. It is an example of what the YAML file looks like. Content in this file is copied from the crdJson JSON variable.

File: solution1/solution1.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		log.SetPrefix("E ")
+		log.Printf("%s", err)
+		os.Exit(1)
+	}
+}
+
+func parseFlags() Config {
+	var config Config
+
+	var homedir string
```
Comment (rossf7) [inline solution1/solution1.go]: Nitpick but I prefer homeDir.

File: solution1/solution1.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+	if err != nil {
+		return err
+	}
+
+	// Create Custom Resource Definition.
+	{
+		log.Printf("creating custom resource")
+
+		res, err := k8sClient.Post(
+			config.K8sServer+"/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions",
```
Comment (rossf7) [inline solution1/solution1.go]: A constant here would be good. At the end we can show how operatorkit helps with paths.

File: solution1/solution1.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+		if err != nil {
+			return err
+		}
+		defer res.Body.Close()
+
+		switch res.StatusCode {
+		case http.StatusConflict:
+			m := make(map[string]interface{})
+			err := json.NewDecoder(res.Body).Decode(&m)
+			return fmt.Errorf("creating custom resource: decoding body=%s", readerToString(res.Body))
```
Comment (rossf7) [inline solution1/solution1.go]: Can the log say it's creating the resource definition?  I can see newbies getting confused between CRDs and CROs at the start.

File: solution1/solution1.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+			minikubeIP := strings.TrimSpace(string(out))
+			serverDefault = "https://" + string(minikubeIP) + ":8443"
+		}
+	}
+
+	flag.StringVar(&config.K8sServer, "kubernetes.server", serverDefault, "Kubernetes API server address.")
+	flag.StringVar(&config.K8sCrtFile, "kubernetes.crt", path.Join(homedir, ".minikube/apiserver.crt"), "Kubernetes certificate file path.")
+	flag.StringVar(&config.K8sKeyFile, "kubernetes.key", path.Join(homedir, ".minikube/apiserver.key"), "Kubernetes key file path.")
+	flag.StringVar(&config.K8sCAFile, "kubernetes.ca", path.Join(homedir, ".minikube/ca.crt"), "Kubernetes CA file path.")
+	flag.Parse()
```
Comment (rossf7) [inline solution1/solution1.go]: I like how this uses the minikube config.

**Review by rossf7 — COMMENTED** [review]
I like the direction here. A few nitpicks but nothing major.

**Comment by kopiczko** [comment]
@rossf7 I think the solution for ex1 is ready. Last bit is to implement `MySQLOps`.
I haven't addressed your comments yet, but I'll do! Thanks for leaving them!

File: solution1/solution1.go [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+
+- two options incluster and remote; remote is useful for testing and
+  prototyping
+- with API objects has to be created with JSON format even though kbuectl
+  allows YAML
+
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
```
Comment (kopiczko) [inline solution1/solution1.go]: Note to self: remove comments below. This was my scratch area :D

File: solution1/solution1.go (hunk 3 above)
Comment (kopiczko) [reply solution1/solution1.go]: I agree, but I don't want ppl to jump around, I'd like them to see immediately the path. We shouldn't consider it production grade code. WDYT?

File: solution1/solution1.go (hunk 4 above)
Comment (kopiczko) [reply solution1/solution1.go]: I think it is ok to call it `custom resource` and `custom object`. (1) it's interchangeable with TPR stuff. (2) Below you have a log for waiting for custom resource to be ready, which is something different than waiting for custom resource definition to be ready. Custom resource is something Kubernetes creates as a result of custom resource definition object reconciliation, and we should make that clear. Do you agree?

File: solution1/crd.yaml (hunk 1 above)
Comment (kopiczko) [reply solution1/crd.yaml]: Done.

File: solution1/solution1.go (hunk 2 above)
Comment (kopiczko) [reply solution1/solution1.go]: Done.

File: solution1/solution1.go (hunk 6 above)
Comment (kopiczko) [reply solution1/solution1.go]: Done.

File: solution1/solution1.go (hunk 3 above)
Comment (rossf7) [reply solution1/solution1.go]: Good point, no need for a constant.

File: solution1/solution1.go (hunk 4 above)
Comment (rossf7) [reply solution1/solution1.go]: Yes, I like custom resource and custom object. Nice that it's interchangeable with TPRs.

File: solution1/solution1.go [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+			log.Printf("checking custom resource readiness attempt=%d: not ready yet", attempt)
+			time.Sleep(checkInterval)
+		}
+	}
+
+	// Start reconciliation loop.
+	reconciliationCnt := 1
+	reconciliationInterval := time.Second * 1
+	for ; ; reconciliationCnt++ {
```
Comment (rossf7) [inline solution1/solution1.go]: Nitpick but the comment here could say it's an infinite loop and that its intended.

**Review by rossf7 — APPROVED** [review]
<3

File: postgresqlops/postgresqlops.go [hunk 8]
```diff
@@ -109,7 +117,7 @@ func (p *PostgreSQLOps) DeleteDatabase(name string) error {
 func (p *PostgreSQLOps) ListDatabases() ([]Database, error) {
 	dbs := []Database{}
 
-	rows, err := p.db.Query("SELECT pg_database.datname, pg_user.usename FROM pg_database, pg_user WHERE pg_database.datdba = pg_user.usesysid")
+	rows, err := p.db.Query("SELECT pg_database.datname, pg_user.usename FROM pg_database, pg_user WHERE pg_database.datdba = pg_user.usesysid AND pg_database.datname NOT IN ('postgres', 'template0', 'template1')")
```
Comment (kopiczko) [inline postgresqlops/postgresqlops.go]: I added some table filtering.

File: postgresqlops/postgresqlops.go [hunk 9]
```diff
@@ -42,6 +45,11 @@ func New(config Config) (*PostgreSQLOps, error) {
 	return postgreSQLOps, nil
 }
 
+// Close relases all PosgreSQLOps resources.
+func (p *PostgreSQLOps) Close() error {
+	return p.db.Close()
+}
```
Comment (kopiczko) [inline postgresqlops/postgresqlops.go]: Added close function.

File: postgresqlops/postgresqlops.go [hunk 10]
```diff
@@ -28,7 +31,7 @@ type PostgreSQLOps struct {
 // New creates the connection to the database.
 func New(config Config) (*PostgreSQLOps, error) {
 	// Postgres user and password are hardcoded and match the resources in postgres.yaml.
-	psqlInfo := fmt.Sprintf("host=%s port=%d user=%s password=%s sslmode=disable", config.Host, config.Port, "postgres", "operator-workshop")
+	psqlInfo := fmt.Sprintf("host=%s port=%d user=%s password=%s sslmode=disable", config.Host, config.Port, config.User, config.Password)
```
Comment (kopiczko) [inline postgresqlops/postgresqlops.go]: User/passwd moved from here.

File: solution1/solution1.go [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/giantswarm/operator-workshop/postgresqlops"
+)
```
Comment (kopiczko) [inline solution1/solution1.go]: I split into multiple files, so this has only interesting code, and not e.g. os signal handling.

File: postgresqlops/postgresqlops.go (hunk 8 above)
Comment (rossf7) [reply postgresqlops/postgresqlops.go]: Good catch! That would cause problems if they were returned.

File: solution1/crd.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# 
+# NOTE:  This file is not used in the solution. It is an example of what the
+# YAML file looks like. Content in this file is copied from the crdJson JSON
+# variable.
+#
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  # name must match the spec fields below, and be in the form: <plural>.<group>
+  name: mysqlconfigs.containerconf.de
```
Comment (rossf7) [inline solution1/crd.yaml]: Can you update the CRD example for PostgreSQL.

**Review by rossf7 — APPROVED** [review]
Thanks, changes are nice.  But the example CRD is still using MySQL.

**Pull request merged**